### ISSUE TITLE
Proposal: return an `isValid` flag in the response to simplify calling code

### ISSUE THREAD

**Issue by fiznool** [AfterShip/phone#246]
Firstly I'd like to say a huge thanks for this library. We've been using it for years in production to parse user-supplied mobile phone numbers into E.164 format, and it has worked beautifully well. ❤️ 

I noticed that the API response has changed for v3, which is great - it's much nicer to work with an object as the response, rather than a tuple. That being said, I was hoping to propose a further improvement, based on an assumption that I'm making (which of course could be wrong).

Right now the response object takes the following format:

``` ts
{
    phoneNumber: string | null;
    countryIso2: string | null;
    countryIso3: string | null;
    countryCode: string | null;
}
```

If I want to ensure that I always return a valid phone number _and_ country ISO code, I have to do the following:

``` ts
import phone from 'phone';

interface ParsedPhoneNumber {
  phoneNumber: string;
  countryIso2: string;
  countryIso3: string;
}

function parsePhoneNumber(phoneNumberInput: string): ParsedPhoneNumber | null {
  const { phoneNumber, countryIso2, countryIso3 } = phone(phoneNumberInput);

  // Explicitly check every field for null, so that TypeScript can infer that each is of type `string` below
  if(phoneNumber === null || countryIso2 === null || countryIso3 === null) {
    return null;
  }

  // At this point, TS has inferred that all three fields are of type `string` - i.e. the `null` types have been erased.
  // We can therefore return them all, satisfying the `ParsedPhoneNumber` return type.
  return {
    phoneNumber,
    countryIso2,
    countryIso3,
  };
}
```

However, if the following assumption is true, this can be simplified via the use of a discriminator field. The assumption is that either:

- All of these properties will be strings, or
- All of these properties will be null

Is that correct? If so, we can use TypeScript's discriminated union support to simplify the above. Consider the following types:

``` ts
interface PhoneNumberInvalidResponse {
  isValid: false;
}

interface PhoneNumberValidResponse {
  isValid: true;
  phoneNumber: string;
  countryIso2: string;
  countryIso3: string;
}

type PhoneNumberResponse = PhoneNumberInvalidResponse | PhoneNumberValidResponse; 
```

Now, if the `phone` library could return the type `PhoneNumberResponse`, i.e. either the valid or the invalid response, we could do the following in code:

``` ts
import phone from 'phone';

interface ParsedPhoneNumber {
  phoneNumber: string;
  countryIso2: string;
  countryIso3: string;
}

function parsePhoneNumber(phoneNumberInput: string): ParsedPhoneNumber | null {
  const parsed = phone(phoneNumberInput);

  // We only need to check the `isValid` field, since this is a discriminator.
  if(!parsed.isValid) {
    return null;
  }

  // At this point, TS knows that `parsed.isValid` must be `true`.
  // It can therefore infer that the type of `parsed` must be `PhoneNumberValidResponse`,
  // and so pulling out these fields will guarantee they are of type `string`.
  // This means we don't need to explicitly check that each field is not `null`.
  const { phoneNumber, countryIso2, countryIso3 } = parsed;
  return {
    phoneNumber,
    countryIso2,
    countryIso3,
  };
}
```

Of course, this all depends on the assumption that if a phone number is valid, it also has the associated country codes returned with it. If this is the case, then using a discriminated union could arguably improve the data model, since it is currently acceptible (according to the return type) for `countryIso2` to be a string and `countryIso3` to be `null`.

**Comment by Bossa573** [comment]
Hello,

Thanks for the support, and yes, returning an array is just because of some sort of historical reason and it did not make sense. 

And the reason it returns an object with null for all the keys with the keys unchanged, 
instead of simply returning a "null",
is because we try to make the interface consistent and reduce the chance that accessing the keys of a "null".

And for your suggestion, I agree that the interface might have a false impression that the user could think that some of the keys can be null. In fact, it's impossible for a response mixing string and null, if one of the key is null, all keys would null, if one of the key got the value, all the keys have the values. So in fact no need to exhaustively check if every keys are null. I would just simply do `if (result.phoneNumber) { ... }`. I think I should update the readme to explain this a little bit.

**Comment by fiznool** [comment]
Thanks for your quick reply!

> So in fact no need to exhaustively check if every keys are null. I would just simply do `if (result.phoneNumber) { ... }`

The problem is TypeScript in this situation - without the exhaustive check, TypeScript still thinks that the properties which haven't been checked are still null.

Taking the example above, if I have a function which expects the following return type:

``` ts
interface ParsedPhoneNumber {
  phoneNumber: string;
  countryIso2: string;
  countryIso3: string;
}
```

The _only_ way to turn the response from `phone` into this type is to perform an exhaustive check. See what happens if we don't:

![phone](https://user-images.githubusercontent.com/820863/125462311-c6b90fd4-0c3a-49d3-a868-051d1efec3c2.png)

If, however, the `phone` library could return a union of types, discriminated by the `isValid` flag, we would know that the response would either include all fields (if `isValid` was `true`), or no fields (if `isValid` is `false`), thus eliminating the need for exhaustive checks.

**Comment by Bossa573** [comment]
I think it indeed raise some usability issue, I'll discuss with the team soon to inspect and improve the interface

**Comment by Bossa573** [comment]
Hello @fiznool , you can check the new version of 3.1.0 for the new interface
https://github.com/AfterShip/phone/releases/tag/3.1.0

**Comment by fiznool** [comment]
Awesome! Thanks very much.

Just wondering if this should have been a major release? I updated to phone 3.1.0 from 3.0.0 and my original code would not compile due to TypeScript errors.

That is, the following code works in v3.0.0:

``` ts
function parsePhoneNumber(phoneNumberInput: string): ParsedPhoneNumber | null {
  const { phoneNumber, countryIso2, countryIso3 } = phone(phoneNumberInput);

  // Explicitly check every field for null, so that TypeScript can infer that each is of type `string` below
  if(phoneNumber === null || countryIso2 === null || countryIso3 === null) {
    return null;
  }

  // At this point, TS has inferred that all three fields are of type `string` - i.e. the `null` types have been erased.
  // We can therefore return them all, satisfying the `ParsedPhoneNumber` return type.
  return {
    phoneNumber,
    countryIso2,
    countryIso3,
  };
}
```

But this fails in v3.1.0:

<img width="670" alt="Screenshot 2021-07-14 at 14 48 45" src="https://user-images.githubusercontent.com/820863/125633362-6c0d01e2-071a-48cf-a27b-8e8e6294f844.png">

**Comment by Bossa573** [comment]
Released 3.1.1 to fix the backward compatibility issue
the issue of 3.1.0 is 
```
interface PhoneNumberInvalidResponse {
  isValid: false;
}

interface PhoneNumberValidResponse {
  isValid: true;
  phoneNumber: string;
  countryIso2: string;
  countryIso3: string;
}
```
basically this interface won't work as expected,
and it makes the object destructuring even more difficult to use under ts environment.

Supposedly no more interface change would be made on 3.x,
I'll close the ticket now, if you got any further proposal,
I guess you can try to create PRs

**Comment by fiznool** [comment]
FWIW I think the original proposal is sound, just that it would need to be released as a v4 given that it is a breaking change.
